### ISSUE TITLE
Synchronize mean/variance of batch normalization when training with multiple GPUs 

### ISSUE THREAD

**Issue by bearpaw** [torch/nn#1071]
Hi everyone,

Recently, I find that when training with multiple GPUs using the `DataParallelTable`, the performance is slightly worse than training with a single GPU. It seems that the problem is caused by that the mean and variance of batch normalization are not synchronized across different GPUs. Does anybody know the solution for this issue? Thank you very much.

**Comment by nhynes** [comment]
I've gotten good use out of code to this effect:
```lua
local function bnparams(self)
   local p, gp, rp, rgp = nn.Module.parameters(self)
    p[#p+1] = self.running_mean
    p[#p+1] = self.running_var
    gp[#gp+1] = self.running_mean:clone():zero()
    gp[#gp+1] = self.running_var:clone():zero()
    return p, gp
end

torch.getconstructortable('nn.BatchNormalization').parameters = bnparams
torch.getconstructortable('cudnn.BatchNormalization').parameters = bnparams
```

That way, the BN params will be distributed with the rest of the params. The only caveat is that you'll have to ensure that you're ignoring those parameters during optimization.

IMHO, these "hidden parameters" lead to enough unexpected results that it might be worth simply including them with the rest of the parameters!

**Comment by bearpaw** [comment]
@nhynes Thanks for the suggestion. However, I don't quite understand what do you mean "you'll have to ensure that you're ignoring those parameters during optimization." Could you please explain more?

**Comment by nhynes** [comment]
You need to make sure that the `gradParameters` are zero for the BN `running_(mean|var)` otherwise your optimizer (e.g. SGD) will unwittingly change them when it takes a gradient step!

Since nothing should be accumulating gradients into the `running_(mean|var)`, an initial call to `zeroGradParameters` should suffice.

**Comment by bearpaw** [comment]
@nhynes Thanks. I use resnet by facebook https://github.com/facebook/fb.resnet.torch . Can I just put the following code in [resnet.lua](https://github.com/facebook/fb.resnet.torch/blob/master/models/resnet.lua) ? Since `zeroGradParameters` would be called before backpropagation in https://github.com/facebook/fb.resnet.torch/blob/master/train.lua#L60.

```lua
local function bnparams(self)
   local p, gp = nn.Module.parameters(self)
    p[#p+1] = self.running_mean
    p[#p+1] = self.running_var
    gp[#gp+1] = self.running_mean:clone():zero()
    gp[#gp+1] = self.running_var:clone():zero()
    return p, gp
end

torch.getconstructortable('nn.BatchNormalization').parameters = bnparams
torch.getconstructortable('cudnn.BatchNormalization').parameters = bnparams
```

**Comment by nhynes** [comment]
Yep, that should do it!

There are two things to note, though:
1. This solution only works during `evaluate` since during `training`, BN uses the statistics of the minibatch. You can sort of work around this by using `nn.BatchNorm` and setting them to `evaluate` after calling `training` on the entire model.
2. Only the `running_(mean|var)` of the first gpu will be preserved. You can sort of work around this by, on all but the first GPU, adding the delta (divided by the number of GPUs, times the learning rate) to the `running_(mean|var)` `gradParameters` so that the first GPU incorporates the changes.

I think that your problem might not be BN specifically, though. Did you significantly increase your batch size? If so, that's likely your problem. Turning up the learning rate helps a bit, but it ends up being an issue of exploration vs exploitation (in terms of the parameter vector on the loss surface). I can link you some papers if you're interested.

**Comment by bearpaw** [comment]
@nhynes Thanks a lot! 

Actually, when I train on 2 GPUs, I use 2 times the batch size than on a single GPU, but the learning rate is kept unchanged. I will try to turn up the learning rate a bit as your suggestions. And it would be also very nice if you can recommend some papers on this issue.

**Comment by nhynes** [comment]
[Here's a paper](https://arxiv.org/abs/1609.04836v1) that describes the effect of batch size on "sharpness" of minima.

I have a technical report that confirms the results and offers some more insight into the problem, but it's not yet posted anywhere and there's yet more work to be done in offering a solution!

**Comment by bearpaw** [comment]
@nhynes I try to increase the learning rate when I increase the batch size. It works fine in my case. Thank you very much.
